### ISSUE TITLE
Add support for variants

### ISSUE THREAD

**Issue by ivarconr** [Unleash/unleash#354]
It would be nice if feature toggles could be extended with a set of variants. We need to  define how the contract changes in order to support this, and how the client should work for boolean values when there also exists variant. We might also want to extend this feature to support some kind of configuration per variant, but I would like us to scope this out for now. 

#297 forms a good starting point. 

#229, #301  are related to this issue.

**Comment by ivarconr** [comment]
My loose thoughts on what we need to consider adding support for variants.  

* Should be possible to add any number of variants to toggle.
    * Minimum requirement is one variant. 
* Toggles with variants should still be possible to use as a boolean-toggle (isEnabled). 
    * Will not get variant information
* Strategies decides whether a toggle is enabled for a given user request. 
    * If the toggle is enabled traffic should be allocated between the variants of the given toggle. 
    * It should be possible to define amount of traffic each variants.
        * How to specify this?
        * Can we start with equal distribution, and implement manual percentage later? The protocol should probably have percentage from beginning, and client SDKs should probably use the percentage. 
* A variant may take extra options (key/value)
* It should be possible to override which variant is served. 
    * Use context params to decide this? 
    * E.g. userId in list should always be served VariantX
* Variant need to have some kind of stickiness for variant allocation
    * (we want same user should always be allocated the same variant)
    * Is this related to overrides?
    * Can we just look for data in the context (if userId exist use that, try sessionId, etc). If we don’t have anything it must be random… :/
    * Applications may chose to store variants for a given user to provide consistency over time. Unleash should probably not deal with this? 
        * You don’t want to remove variants when 
        * You will typically store it in a cookie or similar. 
* Client SDK should provide a new method to deal with variants. 
    * E.g: getVariant(‘toggleName’, context).
    * Should return a variant object. Need to include:
        * Name
        * Options: Map<String, String>. 
        * Metadata? 
            * Overrides.
            * Percentages
    * Could return empty if the case of the toggle is considered disabled for the given user.

**Comment by elhoyos** [comment]
At [AllTheRooms.com](https://www.alltherooms.com/) we've been playing around with variants since the beginning of https://github.com/Unleash/unleash/pull/297 and we feel that https://github.com/alltherooms/unleash-express could fit better under the Unleash GitHub org. It could start serving as an example of cookie persistence for Express.js. If you find this idea interesting, we would love to transfer the ownership.

**Comment by ivarconr** [comment]
> Variants on their own should be independent of the allocation strategy. Someone, could want to allocate based on geolocation and/or device instead of weights. We have special strategies that receive whatever input is needed for the allocation to take place. It all depends on which experimentation strategies you want to use.

This is interesting thoughts, but I fail to see the real use case for this, and feel a this complicates the picture a bit. You typically segment a set of users you want to test a new feature on (activation strategies), and then might want to test a few slight variations on that group. If you want to first want to select only mobile users in sounds like a dedicated feature toggle targeting mobile users and you have a another toggle inside that toggle which is used to test various variants for mobile users. 

I want the first iteration of variants to be super simple and easy to use, and should not depend activation strategies. It might be possible to have custom allocation strategies in the future, but this should they should be defined together with the variants. 

On that note I also planning a new iteration on activation strategies where we rethink the design a bit, where the user can define the strategy from basic building blocks: fields on the context and simple functions (and, or, random, eq) (almost like how kids build programs in (scratch)[https://scratch.mit.edu])

**Comment by ivarconr** [comment]
> At [AllTheRooms.com](https://www.alltherooms.com/) we've been playing around with variants since the beginning of #297 and we feel that https://github.com/alltherooms/unleash-express could fit better under the Unleash GitHub org. It could start serving as an example of cookie persistence for Express.js. If you find this idea interesting, we would love to transfer the ownership.

Yes it would be very interesting to streamlining how to use unleash with context and variants correctly in a real app, and a middle-ware makes sense. If we are able to do it framework independent it would be even better.

**Comment by elhoyos** [comment]
> I fail to see the real use case for this, and feel a this complicates the picture a bit

I was driven by the idea of strategies being both drivers of targeting and segmentation. Here's an example: https://github.com/Unleash/unleash/pull/297#issuecomment-364265462

I see two reasons to leverage both of these decisions under strategies:

1. Nesting toggles: one for target and another for segmentation. It looks like we're transfering the complexity to the user implementing the experiment.
2. Grouping strategies is coming up ahead (https://github.com/Unleash/unleash/issues/229) 🎉 

However I agree, something quite does not fit, specially when as a user you specify the weights in a different location from your variants. Something to think about, I guess. Is that a UI problem?

**Comment by ivarconr** [comment]
@elhoyos I agree that my suggestion on nested toggles was a **poor suggestion** that moves complexity to the user. We should not encourage this! 

> I was driven by the idea of strategies being both drivers of targeting and segmentation. Here's an example: [#297 (comment)](https://github.com/Unleash/unleash/pull/297#issuecomment-364265462)

The example will work out of the box with the current implementation in #379. Strategies is used to drive segmentation and the variant selection will be done using the variant weights. 

> However I agree, something quite does not fit, specially when as a user you specify the weights in a different location from your variants. Something to think about, I guess. Is that a UI problem?

My thinking is that we should make it super simple, easy and intuitive to use Unleash for normal use cases, and I really believe variants with percentages are a typical use case. We also need to make sure that the protocol is intuitive, as there is growing number of clients across platforms.

I think it will be kind of strange if the user should not specify the variant allocation strategy in the same place as she defines variants. And it is not just a UI problem, because the protocol will be looser and make it harder to verify that everything is set up correctly. How to make sure you have spelled your variant name correctly, if it is a strategy Unleash has no knowledge about? 

I also think it is safe to assume that variants needs an override mechanism, where you per variant can define cases where one variant should be preferred. A concrete example here is that you want your own `userId` to get a specific variant in order to test that variant in production. The override mechanism should work on any "context" property, and thus you can use this to make sure that users in a specific country, or all females or whatever you choose to put on the context, will be used to select the variant. 

Based on you feedback it might also be necessary to provide a way to specify custom variant allocation strategies, they might defined the same way you define activation strategies today. However, in the client SDKs they will need to have a different interface as they should return a variant and not just return true/false. 

(One solution could of-course be to reuse activation strategies directly on a variant, but I'm not sure if this makes sense.)

But I suspect must users will not want to implement these strategies themselves. (based on leanings from activation strategies, users feels it is cumbersome to have to implement the code for custom strategies, and in most cases they just want to specify simple rules based on values provided in the context).   

## tldr; 
Anyways, it's really awesome that you care so much, and I love to be challenged on my assumptions and ideas. I will however move forward with the current design in #379 as a hidden beta feature (the user will need to activate variants via a special feature flag, and the implementation is subject to change at any time until we have landed a design we are happy with). We will try it out internally and with selected users of Unleash and gain some real experience in the field before we make final adjustment and make this a permanent feature in Unleash.

**Comment by ivarconr** [comment]
#379
